From 3d0a4b890647e13ea49f34bb4098f5f6cd84dbc1 Mon Sep 17 00:00:00 2001
From: Joris Offouga <offougajoris@gmail.com>
Date: Sun, 15 Mar 2020 20:31:20 +0000
Subject: [PATCH] warp7: add SWUpdate requirements

Signed-off-by: Pierre-Jean Texier <pjtexier@koncepto.io>
Signed-off-by: Joris Offouga <offougajoris@gmail.com>

---
 configs/warp7_defconfig |  7 +++++++
 include/configs/warp7.h | 26 ++++++++++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/configs/warp7_defconfig b/configs/warp7_defconfig
index a5fb48ff1f..47aa574c92 100644
--- a/configs/warp7_defconfig
+++ b/configs/warp7_defconfig
@@ -14,12 +14,14 @@ CONFIG_IMX_HAB=y
 CONFIG_FIT=y
 CONFIG_FIT_VERBOSE=y
 CONFIG_SYS_EXTRA_OPTIONS="IMX_CONFIG=board/warp7/imximage.cfg"
+CONFIG_USE_PREBOOT=y
 CONFIG_BOUNCE_BUFFER=y
 CONFIG_HUSH_PARSER=y
 # CONFIG_CMD_BOOTD is not set
 CONFIG_CMD_BOOTZ=y
 # CONFIG_CMD_IMI is not set
 # CONFIG_CMD_XIMG is not set
+CONFIG_CMD_ENV_FLAGS=y
 CONFIG_CMD_MEMTEST=y
 CONFIG_CMD_DFU=y
 CONFIG_CMD_GPIO=y
@@ -36,8 +38,13 @@ CONFIG_CMD_FAT=y
 CONFIG_CMD_FS_GENERIC=y
 CONFIG_OF_CONTROL=y
 CONFIG_DEFAULT_DEVICE_TREE="imx7s-warp"
+CONFIG_SYS_REDUNDAND_ENVIRONMENT=y
+CONFIG_ENV_OFFSET_REDUND=0xE0000
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
 CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_BOOTCOUNT_LIMIT=y
+CONFIG_BOOTCOUNT_ENV=y
+CONFIG_BOOTCOUNT_BOOTLIMIT=1
 CONFIG_DFU_MMC=y
 CONFIG_DM_I2C=y
 CONFIG_DM_MMC=y
diff --git a/include/configs/warp7.h b/include/configs/warp7.h
index 39c00480bd..1be4d31799 100644
--- a/include/configs/warp7.h
+++ b/include/configs/warp7.h
@@ -27,6 +27,30 @@
 #define CONFIG_DFU_ENV_SETTINGS \
 	"dfu_alt_info=boot raw 0x2 0x1000 mmcpart 1\0" \
 
+#define SWU_ENV_SETTINGS \
+	"bootlimit=1\0"	\
+	"bootcount=0\0"	\
+	"upgrade_available=0\0"	\
+	\
+	"swu_setup="                                                     \
+	"setenv expand_bootargs \"setenv bootargs \\\\\"${bootargs}\\\\\"\"; "              \
+	"run expand_bootargs; "                                             \
+	"setenv expand_bootargs;\0"                                          \
+	\
+	"swu_altbootcmd="	\
+	"if test ${rootpart} = 2; then " \
+	"echo Bountcount limit reached. Reverting to B; " \
+	"setenv rootpart 3; "	\
+	"else "	\
+	"echo Bountcount limit reached. Reverting to A; " \
+	"setenv rootpart 2; "	\
+	"fi; " \
+	"setenv upgrade_available 0; " \
+	"saveenv\0" \
+	\
+	"altbootcmd=run swu_altbootcmd; " \
+	"run bootcmd\0" \
+
 /* When booting with FIT specify the node entry containing boot.scr */
 #if defined(CONFIG_FIT)
 #define BOOT_SCR_STRING "source ${bootscriptaddr}:${bootscr_fitimage_name}\0"
@@ -40,6 +64,8 @@
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	CONFIG_DFU_ENV_SETTINGS \
+	SWU_ENV_SETTINGS \
+	"rootpart=2\0"     \
 	"script=boot.scr\0" \
 	"bootscr_fitimage_name=bootscr\0" \
 	"script_signed=boot.scr.imx-signed\0" \
