From ee50911cc57335b208b48ad7fa8ce2df6557990e Mon Sep 17 00:00:00 2001
From: Pierre-Jean Texier <pjtexier@koncepto.io>
Date: Sun, 10 Mar 2019 20:50:48 +0100
Subject: [PATCH] warp7: add SWUpdate requirements

Signed-off-by: Pierre-Jean Texier <pjtexier@koncepto.io>
---
 configs/warp7_defconfig |  2 ++
 include/configs/warp7.h | 28 ++++++++++++++++++++++++++++
 2 files changed, 30 insertions(+)

diff --git a/configs/warp7_defconfig b/configs/warp7_defconfig
index 955c7af..7810423 100644
--- a/configs/warp7_defconfig
+++ b/configs/warp7_defconfig
@@ -29,6 +29,8 @@ CONFIG_CMD_EXT4_WRITE=y
 CONFIG_CMD_FAT=y
 CONFIG_CMD_FS_GENERIC=y
 CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_BOOTCOUNT_LIMIT=y
+CONFIG_BOOTCOUNT_ENV=y
 CONFIG_DFU_MMC=y
 CONFIG_FSL_ESDHC=y
 CONFIG_OPTEE=y
diff --git a/include/configs/warp7.h b/include/configs/warp7.h
index a391dfb..0f7ba5e 100644
--- a/include/configs/warp7.h
+++ b/include/configs/warp7.h
@@ -41,8 +41,34 @@
 #define CONFIG_DFU_ENV_SETTINGS \
 	"dfu_alt_info=boot raw 0x2 0x400 mmcpart 1\0" \
 
+#define SWU_ENV_SETTINGS \
+	"bootlimit=1\0"	\
+	"bootcount=0\0"	\
+	"upgrade_available=0\0"	\
+	\
+	"swu_setup="                                                     \
+	"setenv expand_bootargs \"setenv bootargs \\\\\"${bootargs}\\\\\"\"; "              \
+	"run expand_bootargs; "                                             \
+	"setenv expand_bootargs;\0"                                          \
+	\
+	"swu_altbootcmd="	\
+	"if test ${rootpart} = 2; then " \
+	"echo Bountcount limit reached. Reverting to B; " \
+	"setenv rootpart 3; "	\
+	"else "	\
+	"echo Bountcount limit reached. Reverting to A; " \
+	"setenv rootpart 2; "	\
+	"fi; " \
+	"setenv upgrade_available 0; " \
+	"saveenv\0" \
+	\
+	"altbootcmd=run swu_altbootcmd; " \
+	"run bootcmd\0" \
+
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	CONFIG_DFU_ENV_SETTINGS \
+	SWU_ENV_SETTINGS \
+	"rootpart=2\0"	\
 	"script=boot.scr\0" \
 	"script_signed=boot.scr.imx-signed\0" \
 	"image=zImage\0" \
@@ -141,6 +167,8 @@
 #define CONFIG_ENV_SIZE			SZ_8K
 
 #define CONFIG_ENV_OFFSET		(8 * SZ_64K)
+#define CONFIG_ENV_OFFSET_REDUND	(10 * SZ_64K)
+
 #define CONFIG_SYS_FSL_USDHC_NUM	1
 
 #define CONFIG_SYS_MMC_ENV_DEV		0
-- 
2.7.4

